### PR TITLE
[examples/suite-classification] Toward resurrecting the classificatio…

### PR DESCRIPTION
**Pull request by harold**
…n example.

 - General project cleanup (delete unused files, etc.)
 - Update deps
 - Rewrite core class in terms of new dataset representation (a seq of maps!)
 - Start to rearrange things towards having experiment be the reusable library---more to do here
 - Some updates to util functions around softmax.

### CODE REVIEW

**Comment by harold** [comment]
@cnuernber --- This doesn't implement everything we discussed to day. But it goes quite far (the example compiles now, and runs under some of the scenarios).

I'd like to merge this, then enumerate and prioritize remaining work as discussed with @rosejn today.

**Comment by harold** [comment]
Marking this as [WIP]. 

Three tasks I'll address before we merge this:
 - [ ] Rename suite-classification to mnist-classification
 - [ ] make the experiment/classification api use datasets instead of folders of folders of images
 - [ ] report classification accuracy time series in the webapp

**Comment by harold** [comment]
@cnuernber @charlesg3 @rosejn --- I am happy with this now.

File: examples/mnist-classification/project.clj [hunk 1]
```diff
@@ -0,0 +1,20 @@
+(defproject mnist-classification "0.9.4-SNAPSHOT"
+  :description "An example of using experiment/classification on mnist."
+  :dependencies [[org.clojure/clojure "1.8.0"]
+                 [thinktopic/experiment "0.9.4-SNAPSHOT"]
```
Comment (charlesg3) [inline examples/mnist-classification/project.clj]: Lets not depend on a snapshot version in examples so people can run the examples with the code thats coming from clojars.

File: examples/mnist-classification/project.clj [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+                 [org.clojure/tools.cli "0.3.5"]
+                 [thinktopic/think.image "0.4.8"]
+                 ;;If you need cuda 8...
+                 [org.bytedeco.javacpp-presets/cuda "8.0-1.2"]
```
Comment (charlesg3) [inline examples/mnist-classification/project.clj]: Don't we default to cuda8? Seems cleaner to remove these lines at this point.

File: examples/mnist-classification/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+Ensuring image data is built, and available on disk.
+Building dataset from folder: mnist/test
+{:answer 2, :guess 2}
+```
+
+A randomly selected handwritten digit from the test dataset is chosen, shown in a popup, and classified using the last exported network. Notably, the `label-one` function only uses functionality from base `cortex`, and none from the `experiment` framework. `cortex` as a library has many fewer dependencies, which simplifies deployment of systems that only need inference without the complexities of the experiment training machinery.
+
+## `train.sh`
+
+The included `train.sh` shell script uses a Makefile and java to run the entire process outside of Lein. It starts a lot faster, runs a bit faster, and hopefully hints at a more repeatable process.
```
Comment (charlesg3) [inline examples/mnist-classification/README.md]: Why not just use `train.sh` to build things rather than have a shell script _and_ a `Makefile`?

File: examples/mnist-classification/src/mnist_classification/core.clj [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            [cortex.util :as util]
+            [cortex.experiment.util :as experiment-util])
+  (:import [java.io File]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Convolutional neural net description
+(def image-size 28)
+(def num-classes 10)
+
+(def initial-description
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Better to make this a function that takes some default parameters (like perhaps the input size) which makes it more generic for people copying this for other classification situations.

File: examples/mnist-classification/src/mnist_classification/core.clj [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+   (layers/dropout 0.5)
+   (layers/linear num-classes)
+   (layers/softmax :id :labels)])
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get Yann LeCun's mnist dataset and save it as folders of folders of png
+;; files. The top level folders are named 'training' and 'test'. The subfolders
+;; are named with class names, and those folders are filled with images of the
+;; appropriate class.
+(defn- ds-data->png
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Seems preferable to move this to a seprate namespace since the point of this example is to focus more on the classification part of things and less on dealing with this particular dataset. That way this file can be a bit more concise / to-the-point.

File: examples/mnist-classification/src/mnist_classification/core.clj [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+   (fn []
+     (println "Ensuring image data is built, and available on disk.")
+     (dorun (map (partial save-image! (str dataset-folder "training"))
+                 (map-indexed vector training-dataset)))
+     (dorun (map (partial save-image! (str dataset-folder "test"))
+                 (map-indexed vector test-dataset)))
+     :done)))
+
+
+(def max-image-rotation-degrees 25)
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Better to put the magic-number type of defs at the top of the file / section and the defonces following them.

File: examples/mnist-classification/src/mnist_classification/core.clj [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+(defn- image-aug-pipeline
+  "Uses think.image augmentation to vary training inputs."
+  [image]
+  (-> image
+      (image-aug/rotate (- (rand-int (* 2 max-image-rotation-degrees))
+                           max-image-rotation-degrees)
+                        false)
+      (image-aug/inject-noise (* 0.25 (rand)))))
+
+
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Why so many spaces?

File: examples/mnist-classification/src/mnist_classification/core.clj [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+  "Creates a BufferedImage suitable for web display from the raw data
+  that the net expects."
+  [observation]
+  (patch/patch->image observation image-size))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The classification experiment system needs a way to go back and forth from
+;; softmax indexes to string class names.
+(def class-mapping
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: I think it might be easier to specify these two separately, especially since they could be specified as functions instead of entries in a map. Just a thought though.

File: examples/mnist-classification/src/mnist_classification/core.clj [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+;; simpler, since `cortex` has many fewer dependencies than `experiment.`
+(def network-filename
+  (str train/default-network-filestem ".nippy"))
+
+
+(defn label-one
+  "Take an arbitrary test image and label it."
+  []
+  (ensure-images-on-disk!)
+  (let [observation (->> (str dataset-folder "test")
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Something like the following `join-path` function is more robust as it doesn't rely on having a `/` at the end of the directory name, but also doesn't fail if it does (I think?).
```
(defn- join-path
  [& args]
(.toString (apply io/file args)))
```

File: examples/mnist-classification/project.clj (hunk 1 above)
Comment (harold) [inline examples/mnist-classification/project.clj]: @cnuernber --- how does Charles' comment here relate to your script that changes all the version numbers together?

File: examples/mnist-classification/project.clj (hunk 2 above)
Comment (harold) [inline examples/mnist-classification/project.clj]: These are still needed.

File: examples/mnist-classification/README.md (hunk 3 above)
Comment (harold) [inline examples/mnist-classification/README.md]: The Makefile tracks dependencies and only rebuilds the jar when necessary.

File: examples/mnist-classification/src/mnist_classification/main.clj [hunk 10]
```diff
@@ -14,7 +14,8 @@
 
 (defn -main
   [& args]
+  (println "Welcome! Please wait while we compile some Clojure code...")
   (let [argmap (parse-opts args cli-options)]
-    (require 'suite-classification.core)
-    ((resolve 'suite-classification.core/train-forever-uberjar)
+    (require 'mnist-classification.core)
```
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/main.clj]: Why is this done this way rather than AOT-ing +/- requiring it?

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 5 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: :) Nope, this file/project are mnist specific. Note that this PR has changes both in examples/mnist-classification (an app) and experiment/classification (a lib).

File: experiment/project.clj [hunk 11]
```diff
@@ -1,7 +1,46 @@
 (defproject thinktopic/experiment "0.9.4-SNAPSHOT"
-  :description "FIXME: write description"
-  :url "http://example.com/FIXME"
-  :license {:name "Eclipse Public License"
-            :url "http://www.eclipse.org/legal/epl-v10.html"}
+  :description "A higher-level library for performing experiments with cortex."
```
Comment (charlesg3) [inline experiment/project.clj]: experiment seems a little generic when this is so tied to (and a part of) cortex. `cortex-experiment` seems a bit more apt, imho.

File: experiment/project.clj [hunk 12]
```diff
@@ -1,7 +1,46 @@
...
+  :garden {:builds [{:id "dev"
+                     :source-paths ["src"]
+                     :stylesheet css.styles/styles
+                     :compiler {:output-to "resources/public/css/app.css"}}]}
+
+  :cljsbuild {:builds
+              [{:id "dev"
+                :figwheel true
+                :source-paths ["src/cljs/"]
+                :compiler {:main "cortex.suite.classify"
```
Comment (charlesg3) [inline experiment/project.clj]: I thought the name `suite` was out?

File: experiment/src/cljs/cortex/suite/classify.cljs [hunk 13]
```diff
@@ -102,14 +103,26 @@
...
+            [:tr [:td (inc i)] [:td a]]))]])
 
 
 (defn classify-component
   [& args]
   (let [confusion-atom (atom {})
-        dataset-atom (atom {})]
-    (js/setInterval #(confusion-matrix-update confusion-atom dataset-atom) 1000)
+        dataset-atom (atom {})
+        accuracy-atom (atom {})]
```
Comment (charlesg3) [inline experiment/src/cljs/cortex/suite/classify.cljs]: Why are these named with `-atom` instead of our traditional `*` suffix?

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 4 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Nice, will do.

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 6 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Good point.

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 7 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Got'em.

File: src/cortex/util.cljc [hunk 14]
```diff
@@ -58,14 +58,20 @@
 
 
 (defn max-index
+  "Good for finding softmax winners and converting back from one-hot."
```
Comment (charlesg3) [inline src/cortex/util.cljc]: `s/Good for//` superfluous... every function should be "good for" something. Should read as `Given a collection...`

**Comment by charlesg3** [comment]
@harold -- nice work and thanks for doing this. Several comments inlined.

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 8 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: mmm, as it stands now I think they need to be maps... the website calls `keys` on one of them to get the class names for the confusion matrix rows/columns.

I could imagine a system where these were functions, but the user would likely also need to provide a sorted list of class names.

That might be low hanging fruit if the system does need to be extended in that way someday.

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 9 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Did not know that. I think what's there is ok for this simple case. If it got more complex or if there was a problem encountered, I would want to do something more robust for sure.

File: examples/mnist-classification/src/mnist_classification/main.clj (hunk 10 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/main.clj]: Not sure, this was already there. I think this hack makes some REPLs start up a bit faster.

File: examples/mnist-classification/README.md (hunk 3 above)
Comment (charlesg3) [inline examples/mnist-classification/README.md]: The `-nt` flag _could_ be used to see if a file is newer obviating the separate file.

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 5 above)
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Still feel that managing the dataset should be separate from training the network.

File: experiment/project.clj (hunk 12 above)
Comment (harold) [inline experiment/project.clj]: https://github.com/thinktopic/cortex/issues/145

Comment (harold) [inline experiment/project.clj]: Nice, missed this when grepping.

File: experiment/src/cljs/cortex/suite/classify.cljs (hunk 13 above)
Comment (harold) [inline experiment/src/cljs/cortex/suite/classify.cljs]: I was just following what was already there; I believe it's because the larger Clojure community doesn't use `*` the way we do. Instead of appending it to deref-ables, they use it to mean 'helper-of'.

Comment (charlesg3) [inline experiment/src/cljs/cortex/suite/classify.cljs]: In that regard, I haven't seen much of the Clojure community attaching the type to the end of variable names in general.

**Comment by harold** [comment]
Thanks @charlesg3 --- got a lot of your comments.

File: experiment/project.clj (hunk 11 above)
Comment (harold) [inline experiment/project.clj]: https://github.com/thinktopic/cortex/issues/145

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 5 above)
Comment (harold) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Anyone using `experiment/classification` is going to need to do both of managing their dataset and defining their network. This namespace represents the minimum amount of work to realistically consume that part of the library. I like them separated by useful comments (as they are now).

**Comment by charlesg3** [comment]
Cool... o/w LGTM! :+1:

File: examples/mnist-classification/src/mnist_classification/core.clj (hunk 5 above)
Comment (charlesg3) [inline examples/mnist-classification/src/mnist_classification/core.clj]: Right -- but my point is that defining their network is likely to look _very_ similar while managing their dataset will likely look quite different.

File: examples/mnist-classification/README.md (hunk 3 above)
Comment (harold) [inline examples/mnist-classification/README.md]: https://github.com/thinktopic/cortex/issues/146

Comment (harold) [inline examples/mnist-classification/README.md]: A PR to compare would be interesting, once this lands. This one tracks changes to all the source files (or at least that's the intent).

Comment (charlesg3) [inline examples/mnist-classification/README.md]: Ah, in that case it would be suitable to do a `find ...  -newer target/uberjar.jar` (or the like) since the Makefile is already shelling out anyway.

File: examples/mnist-classification/project.clj (hunk 2 above)
Comment (cnuernber) [inline examples/mnist-classification/project.clj]: We do default to cuda 8 at this point.

File: examples/mnist-classification/README.md (hunk 3 above)
Comment (cnuernber) [inline examples/mnist-classification/README.md]: A good design here for makefiles would be ideal; there are several places where we really should be using makefiles (anything with garden for example).  Not using them means that various people waste time debugging spurious issues especially in clojurescript and garden worlds.

File: examples/mnist-classification/src/mnist_classification/main.clj (hunk 10 above)
Comment (cnuernber) [inline examples/mnist-classification/src/mnist_classification/main.clj]: It makes the build/run loop *much* faster for uberjar development while adding virtually no additional time during actual running.  I have honestly started with with all uberjar builds because aot isn't worth the time.
